ruby-engine/unleash_engine_ffi.rb

```diff
@@ -0,0 +1,49 @@
+require 'ffi'
+require 'json'
+
+class UnleashEngine
+  extend FFI::Library
+
+  lib_path = ENV['YGGDRASIL_LIB_PATH']
+  raise 'YGGDRASIL_LIB_PATH env variable not set' if lib_path.nil?
+
+  combined_path = File.join(lib_path, 'libyggdrasilffi.so')
+
+  ffi_lib combined_path
+
+  attach_function :engine_new, [], :pointer
+  attach_function :engine_free, [:pointer], :void
+  attach_function :engine_take_state, [:pointer, :string], :string
+  attach_function :engine_is_enabled, [:pointer, :string, :string], :bool
+  attach_function :engine_get_variant, [:pointer, :string, :string], :pointer
+  attach_function :engine_free_variant_def, [:pointer], :void
+
+  def initialize
+    @engine_state = UnleashEngine.engine_new
+    ObjectSpace.define_finalizer(self, self.class.finalize(@engine_state))
+  end
+
+  def self.finalize(engine_state)
```

Comment:
This is another great question!

Short answer - because Ruby is weird

Long answer - we want to generate a lazy callback to clean up the engine state pointer and deallocate the memory. This can't be a object method because then we're generating a reference to the very object we're trying to clean up, which can cause the Ruby VM to leak the memory it allocated in its own memory space. So this boils down to the way objects are garbage collected in Ruby

Reply:
In this case it makes the method static. So it allows for what @sighphyre  is saying.